### ISSUE TITLE
Flow-based infra management

### ISSUE THREAD

**Issue by tj** [apex/apex#470]
This would be primarily a separate project, but it would be integrated into apex(1). I would love some feedback!

## Problem

It's currently really cumbersome and error-prone to set up permissions between sources and functions, you have almost no choice but to go read up on documentation to make sure you're getting it right etc. This is especially annoying with scheduled events.

## Solution

Permissions and resources could be inferred by the graph. For example suppose you had a simple alert processor pipeline with SNS sitting inbetween each set of functions, currently you need about __160__ lines of Terraform and probably even more CloudFormation to do what is effectively:

```
alert_queuer -> sns(topic: alerts) -> alert_processor -> sns(topic: alert_results) -> alert_reporter 
```

The DSL isn't important here but having the ability to view the graph would be pretty intuitive, especially for more complex pipelines.

## Future

- Integration with noflo or similar for graphical FBP would be interesting.

## Issues

- My current apps aren't really complex enough to warrant building all this, but I do think it would be cool
- Need to define the FBP DSL, noflow uses the following but I don't think it would be ideal https://github.com/flowbased/fbp#readme
- `DSL` -> `JSON` conversion
- `JSON` -> {Terraform, CloudFormation}

**Comment by ngmiller** [comment]
I would be extremely interested in support for this kind of abstraction. We're currently using CloudFormation to provision our lambda functions, sns topics, and roles, and using apex to deploy the functions after creation, which has a few quirks. It works reasonably well but we have to ensure the common configuration between CFN and apex is maintained. On top of that, for a pipeline with five events and four lambda functions, we have about 230 lines of CFN template with healthy amounts of copy-pasta.

If you're willing to share, I'd like to hear more of your thoughts on the intermediary JSON format, and what that translation process could look like.

**Comment by ngmiller** [comment]
Oh, for what it's worth, CFN does support previewing changes via change sets: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-changesets.html

**Comment by tj** [comment]
@ngmiller interesting! Maybe that's new? I knew lots of people were complaining about a lack of dry-run before, that's awesome though I'll have to give it a closer look, supporting both somehow would be great.

I haven't thought about the intermediate representation much yet, open to suggestions.

**Comment by DavidTPate** [comment]
Yeah, looks like it is fairly new based on this [blog post](https://aws.amazon.com/blogs/aws/new-change-sets-for-aws-cloudformation/). 

My only concern with CloudFormation is with how it handles syntactically valid (but semantically invalid) updates (for example, specifying an AZ you don't have access to) as those will rollback your entire stack. The use of change sets might prove to help alleviate that (haven't used it much yet).

There's definitely better work that could be done in this space as Terraform is still quite verbose when dealing with repeated things. I've taken to just maintaining some JS files that generate all of my Terraform for me so I don't have to repeat myself so much.

**Comment by DavidTPate** [comment]
Maybe this is a space where some sort of templates offered by Apex could help. They could be coded as Terraform modules and then utilized in that respect to build up the infrastructure. (CloudFormation could do this as well with the input parameters that it allows)

**Comment by tj** [comment]
Yeah I agree, Terraform's modules are currently not powerful enough to help out much there. I would love to have an alternative that didn't require the external dependency on Terraform, but having a DSL like this is probably a decent first step.

**Comment by DavidTPate** [comment]
Agreed, a DSL would be a good start and allow us to support other tools later if desired via plugins (ie. Boto instead of Terraform).
